src/main.rs

```diff
@@ -3233,6 +3252,70 @@ impl RicCall {
             let net_access_points = &main_json[user_id]["NetAccessPoints"];
 
             json["NetAccessPoints"] = (*net_access_points).clone();
+            Ok((jsonobj_to_strret(json, req_id), StatusCode::OK))
+        },
+        RicCall::LinkNic => {
+            check_aksk_auth!(auth);
+            let in_json = require_in_json!(bytes);
+            println!("{:#}", in_json.dump());
+
+            let device_number = require_arg!(in_json, "DeviceNumber").as_i32().unwrap();
+            let nic_id = require_arg!(in_json, "NicId");
+            let vm_id = require_arg!(in_json, "VmId");
+            let (vm_idx, vm) = match get_by_id!("Vms", "VmId", vm_id) {
+                Ok((_, idx)) => (idx, &main_json[user_id]["Vms"][idx]),
+                _ => return bad_argument(req_id, json, "Vm not found")
+            };
+            if vm["State"] != "running" && vm["State"] != "stopped" {
+                return bad_argument(req_id, json, "the VM should be either `running` or `stopped`")
+            }
+            if vm.has_key("Nics") {
+                return bad_argument(req_id, json, "the VM is already linked to a Nic")
+            }
+            if !(1..=7).contains(&device_number) {
+                return bad_argument(req_id, json, "the device number should be between 1 and 7, both included")
+            }
+           
+            let nic = match get_by_id!("Nics", "NicId", nic_id) {
+                Ok((_, idx)) => &mut main_json[user_id]["Nics"][idx],
+                _ => return bad_argument(req_id, json, "Nic not found")
+            };
+            link_nic!(nic, device_number);
+            
+            json["LinkNicId"] = nic["LinkNic"]["LinkNicId"].clone();
+            main_json[user_id]["Vms"][vm_idx]["Nics"] = json::array!{ nic.clone() };
+            Ok((jsonobj_to_strret(json, req_id), StatusCode::OK))
+        },
+        RicCall::UnlinkNic => {
+            check_aksk_auth!(auth);
+            let in_json = require_in_json!(bytes);
+            println!("{:#}", in_json.dump());
+
+            let linknic_id = require_arg!(in_json, "LinkNicId");
+            let nic_idx = match main_json[user_id]["Nics"].members_mut().position(|nic| 
```

Comment:
I think you can use get_by_idx! here.

Reply:
i'm not able to use get_by_id because the macro only takes strs to identify the resource, my comparison is for the LinkNicId which is inside the LinkNic object